base/abstractarray.jl

```diff
@@ -2506,6 +2506,255 @@ end
+ 2.0  4.0  6.0
+
+julia> A = rand(3, 7, 11); E = eachslice(A, dims=2);
+
+julia> (element = size(first(E)), container = size(E))
+(element = (3, 11), container = (7,))
+
+julia> stack(E) == cat(E...; dims=3)
+true
+
+julia> stack(E; dims=2) == A  # inverse of eachslice
+true
+
+julia> M = (fill(i,2,3).+rand.() for i in 1:5, j in 1:7);
+
+julia> (element = size(first(M)), container = size(M))
+(element = (2, 3), container = (5, 7))
+
+julia> stack(M) |> size  # keeps all dimensions
+(2, 3, 5, 7)
+
+julia> stack(M; dims=1) |> size  # vec(container) along dims=1
+(35, 2, 3)
+
+julia> hvcat(5, M...) |> size  # hvcat puts matrices next to each other
+(14, 15)
+```
+"""
+stack(iter; dims=:) = _stack(dims, iter)
+
+"""
+    stack(f, args...; [dims])
+
+Apply a function to each element of a collection, and `stack` the result.
+Or to several collections, [`zip`](@ref)ped together.
+
+The function should return arrays (or tuples, or other iterators) all of the same size.
+These become slices of the result, each separated along `dims` (if given) or by default
+along the last dimensions.
+
+See also [`mapslices`](@ref), [`eachcol`](@ref).
+
+# Examples
+```jldoctest
+julia> stack(c -> (c, c-32), "julia")
+2×5 Matrix{Char}:
+ 'j'  'u'  'l'  'i'  'a'
+ 'J'  'U'  'L'  'I'  'A'
+
+julia> stack(eachcol([1 2 3; 4 5 6]), eachrow([1 -1; 10 -10; 100 -100]); dims=1) do col, row
+         vcat(col .* row, 0, col ./ row)
+       end
+3×5 Matrix{Float64}:
+   1.0    -4.0  0.0  1.0   -4.0
+  20.0   -50.0  0.0  0.2   -0.5
+ 300.0  -600.0  0.0  0.03  -0.06
+```
+"""
+stack(f, iter; dims=:) = _stack(dims, f(x) for x in iter)
+stack(f, xs, yzs...; dims=:) = _stack(dims, f(xy...) for xy in zip(xs, yzs...))
+
+_stack(dims::Union{Integer, Colon}, iter) = _stack(dims, IteratorSize(iter), iter)
+
+# Iterating over an unknown length via append! is slower than collecting:
+_stack(dims, ::IteratorSize, iter) = _stack(dims, collect(iter))
+
+function _stack(dims, ::Union{HasShape, HasLength}, iter)
+    S = @default_eltype iter
+    T = S != Union{} ? eltype(S) : Any  # Union{} occurs for e.g. stack(1,2), postpone the error
+    if isconcretetype(T)
+        _typed_stack(dims, T, S, iter)
+    else  # Need to look inside, but shouldn't run an expensive iterator twice:
+        array = iter isa Union{Tuple, AbstractArray} ? iter : collect(iter)
+        isempty(array) && return _empty_stack(dims, T, S, iter)
+        T2 = mapreduce(eltype, promote_type, array)
+        # stack(Any[[1,2], [3,4]]) is fine, but stack([Any[1,2], [3,4]]) isa Matrix{Any}
+        _typed_stack(dims, T2, eltype(array), array)
+    end
+end
+
+function _typed_stack(::Colon, ::Type{T}, ::Type{S}, A, Aax=_axes(A)) where {T, S}
+    xit = iterate(A)
+    nothing === xit && return _empty_stack(:, T, S, A)
+    x1, _ = xit
+    ax1 = _axes(x1)
+    B = similar(_prototype(x1, A), T, ax1..., Aax...)
+    off = 1
+    if S <: NTuple{<:Any,T} && isbitstype(T) && isbitstype(S)
+        C = reinterpret(reshape, S, B)
```

Comment:
It seems better to accelerate `copyto!(dest, offset, x)` for `x` we known length instead of optimizing `S <: NTuple{<:Any,T}` here.

Reply:
That would be better, I agree. I did not try.

I think `stack(f, xs) = reinterpret(T, f.(xs))` is even quicker for tuples, but returns a weird array type. Maybe a related question is: Should this function be guaranteed to return a new array, or can it be a view? `stack([(1,2), (3,4)])` could also just call `reinterpret`, and `stack(::EachCol)` (or whatever it ends up named) could just call `parent`.

Reply:
OK, I've removed this case here. With https://github.com/JuliaLang/julia/commit/8ad4716e6039deb2749a702fbac210861e100f9d I get the same speed, will make an independent PR.

Thanks!